### PR TITLE
Update Resource Hacker

### PR DESCRIPTION
**Pull request by MalwareMechanic**
Update Resource Hacker to use the latest chocolatey package 5.2.4 (https://community.chocolatey.org/packages/reshack.portable).

Closes https://github.com/mandiant/VM-Packages/issues/739

### CODE REVIEW

File: packages/resourcehacker.vm/resourcehacker.vm.nuspec
Comment (mr-tz) [inline packages/resourcehacker.vm/resourcehacker.vm.nuspec:10]: prefer this or https://community.chocolatey.org/packages/reshack.portable?

Comment (MalwareMechanic) [reply packages/resourcehacker.vm/resourcehacker.vm.nuspec:10]: Oh I missed this, good catch! I'll update b/c I'd prefer the portable version

**Comment by MalwareMechanic** [comment]
It's failing on `windows-2022` with:

```
Attempt to get headers for http://www.angusj.com/resourcehacker/resource_hacker.zip failed.
  The remote file either doesn't exist, is unauthorized, or is forbidden for url 'http://www.angusj.com/resourcehacker/resource_hacker.zip'. Exception calling "GetResponse" with "0" argument(s): "The remote server returned an error: (403) Forbidden."
Downloading reshack.portable 
  from 'http://www.angusj.com/resourcehacker/resource_hacker.zip'
ERROR: The remote file either doesn't exist, is unauthorized, or is forbidden for url 'http://www.angusj.com/resourcehacker/resource_hacker.zip'. Exception calling "GetResponse" with "0" argument(s): "The remote server returned an error: (403) Forbidden." 
This package is likely not broken for licensed users - see https://docs.chocolatey.org/en-us/features/private-cdn.
The upgrade of reshack.portable was NOT successful.
```

Not sure how to handle cases like this since it is out of our control.

@mandiant/flare-vm thoughts?

**Comment by Ana06** [comment]
@MalwareMechanic the issue is that GH is blocking this url, right?

**Comment by Ana06** [comment]
did reshack work?

**Comment by MalwareMechanic** [comment]
Previously the issue was that the URL was inaccessible from the GitHub's perspective (unauthorized or forbidden). In the most recent run, the hash has already changed.

The issue now is we rely on a fixed version of the community package and the community package uses a URL that is NOT versioned (see: https://community.chocolatey.org/packages/reshack.portable#files).

This results in multiple failure points when the hash changes:

- The community package will break if not updated very quickly
- Our package will fail since we rely on an older version of the community package (the hash won't match in the older version due to the URL pointing to a newer tool version).

Below is the most recent error:

```
reshack.portable v5.2.4.386 (forced) [Approved]
reshack.portable package files upgrade completed. Performing other installation steps.
WARNING: Url has SSL/TLS available, switching to HTTPS for download
File appears to be downloaded already. Verifying with package checksum to determine if it needs to be redownloaded.
Error - hashes do not match. Actual value was 'F958DB1D239E69051145[77](https://github.com/mandiant/VM-Packages/actions/runs/6828057488/job/18937701487?pr=741#step:4:78)7DE9943B267A3230CC3D140599B48CF024E2C8B3A2'.
Downloading reshack.portable 
  from 'https://www.angusj.com/resourcehacker/resource_hacker.zip'

Download of resource_hacker.zip (3.2 MB) completed.
Error - hashes do not match. Actual value was 'F958DB1D239E69051145777DE9943B267A3230CC3D140599B48CF024E2C8B3A2'.
ERROR: Checksum for 'C:\Users\runneradmin\AppData\Local\Temp\chocolatey\reshack.portable\5.2.4.386\resource_hacker.zip' did not meet '8e56a5c84999f036355759e5cb759e3055df761b55d9ae6a72[82](https://github.com/mandiant/VM-Packages/actions/runs/6828057488/job/18937701487?pr=741#step:4:83)5d4199f328e9' for checksum type 'sha256'. Consider passing the actual checksums through with --checksum --checksum64 once you validate the checksums are appropriate. A less secure option is to pass --ignore-checksums if necessary.
```

Since the download for the tool isn't versioned: https://angusj.com/resourcehacker/#download, we'll continue to run into hash issues even if we try to maintain a custom package. Though if we maintained a custom package we may be able to bypass hash checks if we really needed to (though off hand I'm not familiar with how to do this).

**Comment by MalwareMechanic** [comment]
A stop gap we could do is not version the dependency in our metapacakge and hope that the community package is updated when the hash changes.

**Comment by MalwareMechanic** [comment]
So strange, now it's trying to use `resourcehacker.portable` instead of `reshack.portable`:

```
Upgrading the following packages:
resourcehacker.vm
By upgrading, you accept licenses for the packages.

resourcehacker.portable v5.1.8 (forced) [Approved]
resourcehacker.portable package files upgrade completed. Performing other installation steps.
WARNING: Url has SSL/TLS available, switching to HTTPS for download
Downloading resourcehacker.portable 
  from 'https://www.angusj.com/resourcehacker/resource_hacker.zip'

Download of resource_hacker.zip (3.2 MB) completed.
Error - hashes do not match. Actual value was 'F958DB1D239E69051145777DE9943B267A32[30](https://github.com/mandiant/VM-Packages/actions/runs/6960518789/job/18940215513?pr=741#step:4:31)CC3D140599B48CF024E2C8B3A2'.
ERROR: Checksum for 'C:\Users\runneradmin\AppData\Local\Temp\chocolatey\resourcehacker.portable\5.1.8\resource_hacker.zip' did not meet 'D158BEBC2993CF6BEBF2C23A9[35](https://github.com/mandiant/VM-Packages/actions/runs/6960518789/job/18940215513?pr=741#step:4:36)72A68544C2BA5AE0565[38](https://github.com/mandiant/VM-Packages/actions/runs/6960518789/job/18940215513?pr=741#step:4:39)F70A58075C9[39](https://github.com/mandiant/VM-Packages/actions/runs/6960518789/job/18940215513?pr=741#step:4:40)2D6' for checksum type 'sha256'. Consider passing the actual checksums through with --checksum --checksum64 once you validate the checksums are appropriate. A less secure option is to pass --ignore-checksums if necessary.
The upgrade of resourcehacker.portable was NOT successful.
```

**Comment by MalwareMechanic** [comment]
Alright, so if I revert our metapackage version the modified `.nuspec` file is not picked up in the change for testing ( test uses the old version). So we're at a standstill until we figure out the direction we want to go.

I see the following solutions:

- Ask the tool author to maintain versioned download links
- Delete and remake our metapackage with versioning syntax 0.0.0.YYYYMMDD and use a non-versioned dependency

Thoughts @Ana06 ?

**Comment by Ana06** [comment]
> Delete and remake our metapackage with versioning syntax 0.0.0.YYYYMMDD and use a non-versioned dependency

We have decided not to do this for other packages because then we are "acepting" having broken packages. This is an issue when installing FLARE-VM as you don't know if some packages will install until you try. It also increases the number of bug reports we get from users. If we go this route, I wouldn't add this package to the default flare-vm configuration, which I think it is a pity because this is a great tool.

So I would prefer to ask the tool author if he is willing to help us to be able to include his tool in the default flare-vm installation. @MalwareMechanic can you contact him? His contact details are in https://www.angusj.com/resourcehacker

**Comment by Ana06** [comment]
Also, if the download url is blocked by GH in Windows 11, the tool maintainer may also be able to help us fix it, either by providing another url or by requesting GH not to block this url (not sure if we could request it).

**Comment by Ana06** [comment]
@MalwareMechanic any updates? 😄

**Comment by Ana06** [comment]
There seems to be a similar issue with peanatomist: https://github.com/mandiant/VM-Packages/issues/831

**Comment by MalwareMechanic** [comment]
@Ana06 My previous attempt at contacting the author wasn't successful, but I'll reach out again and ping back if I hear anything

**Comment by Ana06** [comment]
@MalwareMechanic is the tool signed? If so, we could use `VM-Assert-Signature` to install it.

@MalwareMechanic was reshack working (vs the portable version)? I think it would be ok to use that version if it works.

**Comment by naacbin** [comment]
Both version works but `reshack.portable` seems to update faster than `resourcehacker.portable`. It has been 4 months since any new updates and this package still failed to install, I think we should at least update to the latest version for now and keep an issue for long term support.

**Comment by naacbin** [comment]
> @naacbin I agree. Would you like to take over fixing resourcehacker? @MalwareMechanic is likely not going to have time soon.

I will see to do it before end of week.

**Review by Ana06 — COMMENTED** [review]
Thanks a lot for the changes @naacbin!! I have tested locally and it works. Just some comments about the version/package name

File: packages/resourcehacker.vm/resourcehacker.vm.nuspec
Comment (naacbin) [reply packages/resourcehacker.vm/resourcehacker.vm.nuspec:10]: I haven't specify the version as it will fail if the URL in the metacpakge doesn't change anyway.

File: packages/resourcehacker.vm/resourcehacker.vm.nuspec
Comment (naacbin) [reply packages/resourcehacker.vm/resourcehacker.vm.nuspec:5]: Not a big fan of naming the package `reshack.portable.vm` as I think it's cleaner to have the tool name. I change it to `5.2.7.20240413` because I hit same error as https://github.com/mandiant/VM-Packages/pull/741#issuecomment-1823190837

Comment (naacbin) [reply packages/resourcehacker.vm/resourcehacker.vm.nuspec:5]: Since `resourcehacker.portable` is working too I suggest following nuspec file so that we keep same package name :
```xml
    <id>resourcehacker.vm</id>
    <version>5.2.7</version>
    <dependencies>
      <dependency id="common.vm" />
      <dependency id="resourcehacker.portable" version="[5.2.7]" />
    </dependencies>
```

Comment (Ana06) [reply packages/resourcehacker.vm/resourcehacker.vm.nuspec:5]: there are only two options to follow the documentation:
1. rename the package to `reshack.portable.vm`
2. change the package version to `0.0.0.DATE`

Using `resourcehacker.portable` as dependency, `resourcehacker.vm` as package name, and `5.2.7` as package version **does not follow the documentation**. Using this setting will cause that our bot updates the dependency version, but not the package version (as it is not a metapackage because the name of the dependency doesn't match the package name). We had have this issue with other packages that were not following the documentation (and we need to fix them): https://github.com/mandiant/VM-Packages/issues/985. So, if you want to keep `resourcehacker.vm` as the package name, you need to change the package version to `0.0.0.DATE`.

Comment (naacbin) [reply packages/resourcehacker.vm/resourcehacker.vm.nuspec:5]: As mentioned above, it is not possible to change it to `0.0.0.20240422`. By default the latest version is install based on test script and CI/CD (which is 5.1.8 from your repository `https://www.myget.org/F/vm-packages/api/v2`). If you can remove it from myget it will work but I don't see other solution.

Comment (naacbin) [reply packages/resourcehacker.vm/resourcehacker.vm.nuspec:5]: I have based my code on the example you provide in the comment `cygwin` : https://github.com/mandiant/VM-Packages/blob/main/packages/cygwin.vm/cygwin.vm.nuspec#L5
A change might be required there too

Comment (Ana06) [reply packages/resourcehacker.vm/resourcehacker.vm.nuspec:5]: Sorry I was not clear, I'll delete the previous version from myget after you do the change decreasing the version.

cygwin does not need to be changed. The dependency and the package have the exact same name, so it is considered to be a metapackage. The issue with your changes is that you want to keep two different names and in this case the package is not considered a metapackage and the dependency is updated without updating the package name. As mentioned already, there are only two options to follow the documentation:
1. rename the package to `reshack.portable.vm`
2. change the package version to `0.0.0.DATE`

Comment (naacbin) [reply packages/resourcehacker.vm/resourcehacker.vm.nuspec]: Oh, that's why, thanks for the explanation. I have update the package as request

**Comment by Ana06** [comment]
Thanks for the changes @naacbin! I have removed the old package from MyGet and retriggered the test suite so that the new package is tested.

**Review by Ana06 — APPROVED** [review]
Thanks @naacbin! It works perfectly. I think we should add this package to the default flare-vm configuration. :smile:

**Pull request merged**
